modules/openapi-generator/src/main/java/org/openapitools/codegen/CodegenModel.java

```diff
@@ -345,6 +345,15 @@ public boolean getIsClassnameSanitized() {
         return !StringUtils.equals(classname, name);
     }
 
+    /**
+     *
+     * @return true if @JSonTypeName is needed
+     */
+    public boolean getUseJsonTypeName() {
+        Object xDiscriminatorValue = vendorExtensions.get("x-discriminator-value");
+        return getIsClassnameSanitized() || (xDiscriminatorValue!=null && !classname.equals(xDiscriminatorValue));
+    }
+
```

Comment:
This code block is for Java only so I don't think we should include it in Codegen Model.

What about moving it to Abstract Java Codegen instead?